### PR TITLE
feat(lambda-layer-kubectl): custom lambda layer for kubectl based on EKS version of the cluster

### PR DESCRIPTION
**Pull request by robertd**
This is an early WIP/POC

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [ ] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md)?
	* [ ] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by pahud** [comment]
Awesome!

File: packages/@aws-cdk/lambda-layer-kubectl/lib/kubectl-layer.ts [hunk 1]
```diff
@@ -1,27 +1,32 @@
...
+   */
+  readonly kubectlVersion: string;
+}
+
 /**
  * An AWS Lambda layer that includes `kubectl` and `helm`.
  */
 export class KubectlLayer extends lambda.LayerVersion {
-  constructor(scope: Construct, id: string) {
+  constructor(scope: Construct, id: string, props: KubectlLayerProps) {
```
Comment (pahud) [inline packages/@aws-cdk/lambda-layer-kubectl/lib/kubectl-layer.ts:21]:
```suggestion
  constructor(scope: Construct, id: string, props?: KubectlLayerProps) {
```
We probably should make it optional for backward compatibility?

File: packages/@aws-cdk/lambda-layer-kubectl/test/kubectl-layer.test.ts [hunk 2]
```diff
@@ -7,7 +7,7 @@ test('synthesized to a layer version', () => {
   const stack = new Stack();
 
   // WHEN
-  new KubectlLayer(stack, 'MyLayer');
+  new KubectlLayer(stack, 'MyLayer', { kubectlVersion: '1.22' });
```
Comment (pahud) [inline packages/@aws-cdk/lambda-layer-kubectl/test/kubectl-layer.test.ts:10]:
```suggestion
  new KubectlLayer(stack, 'MyLayer', { kubectlVersion: '1.22.6' });
```

**Review by pahud — CHANGES_REQUESTED** [review]
Good job! Thanks for the PR @robertd !

It's pretty much the same with my experimental code today.  Some suggested changes provided though and some considerations as below:

1.  To avoid breaking changes, maybe we should allow the `props` to be optional with predefined default versions of `kubectl` and `helm` CLIs. However, users are allow to customize them if necessary.

2. We probably need a static mapping table in  aws-eks for `cluster version` with `kubectl version`, e.g.
 cluster 1.22 should map to kubectl 1.22.6 according to [this document] (https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html) so when we provision a 1.22 cluster, the kubectl of version `1.22.6` lambda layer will be provisioned.
 
 Thoughts?

File: packages/@aws-cdk/lambda-layer-kubectl/lib/kubectl-layer.ts (hunk 1 above)
Comment (robertd) [reply packages/@aws-cdk/lambda-layer-kubectl/lib/kubectl-layer.ts:21]: Agreed.

File: packages/@aws-cdk/lambda-layer-kubectl/test/kubectl-layer.test.ts (hunk 2 above)
Comment (robertd) [reply packages/@aws-cdk/lambda-layer-kubectl/test/kubectl-layer.test.ts:10]: Yep... tests will most likely have hardcoded value unless we create a dummy cluster that will expose this piece of information.

File: packages/@aws-cdk/lambda-layer-kubectl/lib/kubectl-layer.ts [hunk 3]
```diff
@@ -1,27 +1,32 @@
...
     super(scope, id, {
-      code: lambda.Code.fromAsset(path.join(__dirname, 'layer.zip'), {
-        // we hash the Dockerfile (it contains the tools versions) because hashing the zip is non-deterministic
-        assetHash: hashFile(path.join(__dirname, '..', 'layer', 'Dockerfile')),
+      code: lambda.Code.fromDockerBuild(path.join(__dirname, '..', 'layer'), {
+        buildArgs: {
+          KUBECTL_VERSION: props.kubectlVersion,
+          HELM_VERSION: '3.8.2',
```
[the remark this answers is not in the document]
Comment (robertd) [reply packages/@aws-cdk/lambda-layer-kubectl/lib/kubectl-layer.ts:26]: Will `KubectlLayer` ever be used outside of Cluster constructor?

**Comment by pahud** [comment]
Have discussed with @neilkuan today about this issue. What we can explore next this to pass the cluster version to here so the `KubectlProvider` might be able to create correct layer.

https://github.com/aws/aws-cdk/blob/9c93ba75fb913d27db79ea9618f3c7af2c665e22/packages/%40aws-cdk/aws-eks/lib/kubectl-provider.ts#L155

**Comment by robertd** [comment]
@pahud ... few notes:

- This PR introduces direct dependency on Docker for building kubectl layer image "on the fly". In the past few months, Docker for Desktop license changed for companies and lots of folks are migrating to alternatives.
- We've observed increase in the duration of tests. For example `aws-eks` tests are now around ~5min (up from 30-60s). "Downstream modules" that have `aws-eks` and `lambda-layer-kubectl` as dependencies will see increase in duration too.
- Tests are still failing. Still investigating.
- We are introducing manual management of EKS version mapping in `packages/@aws-cdk/aws-eks/lib/kubectl-provider.ts`. I really wish kubectl download urls supported `major.minor` in the url and provide the latest version of the `kubectl` - rather than having to specify  the `major.minor.patch` version in the url. This would make a world of difference for us (downstream users) and we wouldn't have to do manual version mapping in `packages/@aws-cdk/aws-eks/lib/kubectl-provider.ts`.

  - `https://storage.googleapis.com/kubernetes-release/release/v1.22/bin/linux/amd64/kubectl` (preferred)
  vs 
  `https://storage.googleapis.com/kubernetes-release/release/v1.22.9/bin/linux/amd64/kubectl` (actual)
- We bumped `helm` to `3.9.x`. I'm not expecting any issues going to `3.9.x`.

**Comment by pahud** [comment]
According to the approach proposed in [RFC 39](https://github.com/aws/aws-cdk-rfcs/pull/430), we are planning to publish `@aws-cdk/lambda-layer-kubectl-X` separately and this will avoid breaking changes in the aws-eks API. For users to deploy eks 1.22 clusters, `@aws-cdk/lambda-layer-kubectl-122` would be required. I think this might be a better approach. wdyt?

**Pull request closed without merging**